### ISSUE TITLE
Autorest PCL Support.

### ISSUE THREAD

**Issue by DavidStrickland0** [Azure/autorest#2584]
The Microsoft.Rest.ClientRuntime. project appears to have dropped PCL support. The last version that supports PCL being 2.3.6. Unfortunately 2.3.6 also restricts the NewtonSoft Version >=9 but <10. 

With Autorest relying on the ClientRuntime and the ClientRuntime PCL support deprecated and locked into a specific version of Newtonsoft. AutoRest no longer supports generating code for use in a PCL unless you want to permenently lock into a sunsetting version of Newtonsoft. Not sure if this is an issue or by intent but couldnt find any other discussions on the topic nor any annoucement that Autorest is deprecating PCL support.

**Comment by olydis** [comment]
@DavidStrickland0 The client runtime and its developers live [here](https://github.com/Azure/azure-sdk-for-net), would you mind reposting the issue there so people with the same question may find it? 🙂 @shahabhijeet FYI

**Comment by DavidStrickland0** [comment]
@Olydis Can do but I'm assuming the Clients answer is that PCL's are deprecated and being replaced. I was just wondering is thats Autorests position as well.

**Comment by olydis** [comment]
Oh I see, I think I parsed your question wrong, sorry! AutoRest itself has no intentions on deprecating anything unless it becomes painful to support. Do you mean AutoRest is generating C# code that is not PCL compatible? We try to be as broad as possible :-) The ClientRuntime should indeed be the only entity with such restriction (since it's assemblies with dependencies and such, not just code). Note that AutoRest doesn't force you to use any specific version of the ClientRuntime - it recommends a recent one, since bugs will be fixed in there, sometimes there's updates in (say) polling/retry behavior, etc., but there shouldn't have been a breaking change in a very very long time, so the choice is essentially up to you! MS internally we have strong guidelines regarding which CR version is used for released Azure SDKs and such, but external customers are essentially welcome to use whichever version works for them together with generated code.

**Comment by DavidStrickland0** [comment]
From my generated APIService
public abstract partial class ApiService : ServiceClient<ApiService>

ServiceClient is a direct Reference to Microsoft.Rest.ClientRuntime.

**Comment by DavidStrickland0** [comment]
But by building AutoRest to inherit from a base class that doesnt support PCL's you've basically stopped Support for PCL Code generatation from Autorest. The fact that the dependency is inherited instead of injected means the AutoRest generated code is restricted to supporting only what the base class supports. Would it not be better to inject the dependency on the ClientRuntime?

**Comment by DavidStrickland0** [comment]
Anyway really just wanted to start the conversation and get the issue out there. Our short-term plan is to downgrade our Newtonsoft version until we are ready to go to .NetStandard. But perhaps just the existence of this "issue" will help others avoid the problem.

**Comment by olydis** [comment]
But the base class you are talking about exists in super old versions of the CR, doesn't it? Are all of those not PCL compatible? Injection of a base class sounds dangerous and overly complicated - even more since you already aren't tied to any specific implementation! AutoRest generates *source code*, so you can compile it with anything you like! You don't even need the CR at all if you put an appropriate class into the same project, so I'm not sure what the benefit would be.

**Comment by DavidStrickland0** [comment]
> But the base class you are talking about exists in super old versions of the CR, doesn't it?

The current version of Autorest generates C# source code that inherits from  

"ServiceClient<ApiService>, IApiService" 

using Microsoft.Rest;
using Microsoft.Rest.Serialization;

Hard references writen in that I would have to change after generation to either use something else or avoid naming conflicts.  Now granted I can change all this every time I generate my client in fact I can write the entire thing from scratch and never use Autorest but thats not really the point.

> "Whatever you expect to be injected you can already just make available during compilation."

In this case its Something I want to REMOVE not add. Namely the hard dependency on a specific version of Newtonsoft while maintaining compatability with PCL. I could write my own ClientRuntime or alter the existing one to do this but I would also have too rework the source code generater to generate code that references my new ClientRuntime instead. If nothing else either changing the name of the base class or the usings. Or Id have to give my new client the same Namespace and ClassName as the existing Microsoft Client so that AutoRest Generated Code could find it. None of these is particularly clean solutions.
  
The ability to inject the Dependency on the ClientRuntime in the _source code_ would truely decouple AutoRest from the Runtime instead of being restricted to only one implementation of the dependencies contract incidentally that likewise goes for the AutoRest Dependency on HTTPClient but discussing the advantages of IOC and Dependency injection over hard references is a bit beyond the scope of this discussion.

**In short** 
Autorest generated code relies on a dependency that no longer supports PCL and theres no easy way to fix this or swap it out with something else that does without making alterrations to the generated code after the fact.

**Comment by olydis** [comment]
I think there must me some kind of big misunderstanding.

> Hard references writen in that I would have to change after generation

No, these are not hard references, just `using` statements. I have actually checked now: CR version 2.2.0 and its dependencies is PCL compatible and works perfectly well with generated code. You do not have to edit a single line of generated code. Why does that not work for you?

> I would also have too rework the source code generater to generate code that references my new ClientRuntime instead

No, the generated code does *not* contain references already! You can use whatever ClientRuntime you want already! References are in *project files* which we don't even generate - we only generate source code which by design of C# uses whatever was referenced from the project file. Fully decoupled concepts already.

> In short
> Autorest generated code relies on a dependency that no longer supports PCL 

no

> and theres no easy way to fix this or swap it out with something else that does without making alterrations to the generated code after the fact.

yes, just compile with CR 2.2.0 or earlier

**Comment by DavidStrickland0** [comment]
> No, these are not hard references, just  using  statements
What qualifies as a hard reference then? Can you make any more concrete tightly coupled reference then a using statement? 

> No, the generated code does not contain references already! You can use whatever ClientRuntime you want already!.....  just compile with CR 2.2.0 or earlier.
I think there must me some kind of big misunderstanding.

> yes, just compile with CR 2.2.0 or earlier
Um that would be 2.3.6 PCL support was dropped in 2.3.7.

>Everything works just use an old version.... 
I think there must me some kind of big misunderstanding.

**Comment by olydis** [comment]
> What qualifies as a hard reference then? Can you make any more concrete tightly coupled reference then a using statement?

The reference you have in the project file (that one is actually called "reference" and points to a specific assembly) - AutoRest does not make any assumption/restriction on what you do there.
Using statements simply import stuff with given name from whatever was referenced by your project.

> Um that would be 2.3.6 PCL support was dropped in 2.3.7.

According to nuget.org it was dropped from version 2.2.0 to 2.3.0 but maybe they display that information wrong.

Yes, I think there is a misunderstanding, what specifically doesn't work in your case if you reference a CR that is still PCL compatible? Does it not compile? What is the error? Please provide us with an issue we can reproduce.
